src/Microsoft.AspNetCore.Mvc.ApiExplorer/ApiBehaviorApiDescriptionProvider.cs

```diff
@@ -50,67 +60,134 @@ public bool AppliesTo(ApiDescription description)
             return description.ActionDescriptor.FilterDescriptors.Any(f => f.Filter is IApiBehaviorMetadata);
         }
 
-        // Check if the parameter is named "id" (e.g. int id) or ends in Id (e.g. personId)
-        public bool IsIdParameter(ParameterDescriptor parameter)
+        public bool HasUserDefinedResponseTypes(ApiDescription description)
+        {
+            return description.ActionDescriptor.FilterDescriptors
+                .Select(f => f.Filter)
+                .OfType<IApiResponseMetadataProvider>()
+                .Any(f => f.Type != null);
+        }
+
+        public void ApplyProfile(ApiDescription description)
         {
-            if (parameter.Name == null)
+            ApiDescriptionProfile profile = null;
+
+            var profiles = _apiOptions.Value.ApiDescriptionProfiles;
+            for (var i = 0; i < profiles.Count; i++)
             {
-                return false;
+                if (profiles[i].IsMatch(description))
+                {
+                    profile = profiles[i];
+                    break;
+                }
             }
 
-            if (string.Equals("id", parameter.Name, StringComparison.Ordinal))
+            if (profile == null)
             {
-                return true;
+                return;
             }
 
-            // We're looking for a name ending with Id, but preceded by a lower case letter. This should match
-            // the normal PascalCase naming conventions.
-            if (parameter.Name.Length >= 3 && 
-                parameter.Name.EndsWith("Id", StringComparison.Ordinal) && 
-                char.IsLower(parameter.Name, parameter.Name.Length - 3))
+            // This is just for debugging/testing purposes.
+            description.SetProperty(profile);
+
+            profile.ApplyTo(description);
+
+            // Now we need to fixup some of the ApiResponseTypes - we don't expect a profile to be able to
+            // apply the content negotiation info, or to have access to ModelMetadata.
+            for (var i = 0; i < description.SupportedResponseTypes.Count; i ++)
             {
-                return true;
-            }
+                var response = description.SupportedResponseTypes[i];
+                if (response.Type == null || response.Type == typeof(void))
+                {
+                    continue;
+                }
+
+                if (response.ModelMetadata == null)
+                {
+                    response.ModelMetadata = _modelMetadaProvider.GetMetadataForType(response.Type);
+                }
 
-            return false;
+                if (response.ApiResponseFormats.Count == 0)
+                {
+                    var formats = GetFormats(description, response);
+                    foreach (var format in formats)
+                    {
+                        response.ApiResponseFormats.Add(format);
+                    }
+                }
+            }
```

Comment:
@pranavkm - this is very similar to some code in the default provider. We might think about putting it behind a service or something. Basically even though all of these things are POCOs, the `ApiResponseFormat` requires a lot of boilerplate to construct correctly.

This is the reason why your code will advertise `application/problem+xml` with no XML formatter configured.

Reply:
This comment actually refers to the `GetFormats` method